src/mono/mono/mini/interp/jiterpreter.c

```diff
@@ -575,7 +575,7 @@ mono_jiterp_interp_entry_prologue (JiterpEntryData *data, void *this_arg)
 	method = rmethod->method;
 	jiterp_assert(method);
 
-	if (mono_interp_is_method_multicastdelegate_invoke(method)) {
+	if (data->header.can_be_invoke && mono_interp_is_method_multicastdelegate_invoke(method)) {
```

Comment:
you've got a byte, is there some reason you can't store the result of the check once it has run too?

Reply:
Great point, I'll do that.

Reply:
Actually, I'm not certain this value is guaranteed to be the same on every call? @vargaz @BrzVlad do either of you know whether the 'is multicastdelegate invoke' state when entering the interp could change from call to call? It seems to me like it should be constant for a given entry wrapper.

Reply:
https://github.com/dotnet/runtime/blob/db28821852151f62fb81d83d0fd1c8de9e0e5640/src/mono/mono/mini/interp/interp.c#L2165 appears to be the logic

Reply:
Aren't interp entry wrappers generated per method ? So we know at jiterp compile time if the method is multicast delegate invoke. Which means we could remove the checks for this altogether from interp entry path and generate a separate call to obtain the delegate wrapper when we know we are calling such a method. Or maybe not bother at all with compiling jit wrappers for these methods. Are they really that common ?

Reply:
My understanding is that they're per-method so we know at jiterp compile time, yeah. I just wasn't 100% certain that I wasn't missing a code path where they could get shared. I don't know how to identify whether they're actually common in production software or not... I suspect they're going to get hit a lot in System.Text.Json, I'll see if I can verify that.

Reply:
I can be stored in InterpMethod.

Reply:
Verified that storing a flag in InterpMethod works and that the S.T.J tests hit this as I expected. So the optimization applies to both regular AOT->interp transitions and to the jiterpreter now. Vlad suggested generating optimized wrappers that skip the check entirely, which I'll look into doing.